// @metamask/snaps-cli/dist/cmds/eval/eval-worker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
// eslint-disable-next-line import/no-unassigned-import
require("ses/lockdown");
const worker_threads_1 = require("worker_threads");
const mock_1 = require("./mock");
lockdown({
    consoleTaming: 'unsafe',
    errorTaming: 'unsafe',
    mathTaming: 'unsafe',
    dateTaming: 'unsafe',
    overrideTaming: 'severe',
});
if (worker_threads_1.parentPort !== null) {
    worker_threads_1.parentPort.on('message', (message) => {
        const { snapFilePath } = message;
        const snapModule = { exports: {} };
        new Compartment({
            ...getMockEndowments(),
            module: snapModule,
            exports: snapModule.exports,
        }).evaluate((0, fs_1.readFileSync)(snapFilePath, 'utf8'));
        if (!snapModule.exports?.onRpcRequest) {
            console.warn("The Snap doesn't have onRpcRequest export defined");
        }
        setTimeout(() => process.exit(0), 1000); // Hack to ensure worker exits
    });
}
function getMockEndowments() {
    const endowments = (0, mock_1.generateMockEndowments)();
    return {
        ...endowments,
        window: endowments,
        self: endowments,
    };
}
